Align arm64 data section as requested

BruceForstall opened the pull request:
Currently, the data section alignment request is ignored unless
it is 8. Since the minimum is 4, this effectively means that
16-byte SIMD16 data alignment requests are ignored. This is likely
because this code was written before arm64 supported SIMD, and was
never revised.

Cases of SIMD loads of constant data lead to larger alignment
padding of the data section. This is somewhat mitigated by
https://github.com/dotnet/runtime/pull/71043 which fixes a bug with overallocation
and overalignment of SIMD8 data loads.

BruceForstall committed:
Align arm64 data section as requested

Currently, the data section alignment request is ignored unless
it is 8. Since the minimum is 4, this effectively means that
16-byte SIMD16 data alignment requests are ignored. This is likely
because this code was written before arm64 supported SIMD, and was
never revised.

Cases of SIMD loads of constant data lead to larger alignment
padding of the data section. This is somewhat mitigated by
https://github.com/dotnet/runtime/pull/71043 which fixes a bug with overallocation
and overalignment of SIMD8 data loads.

BruceForstall commented:
@kunalspathak @tannergooding PTAL
cc @dotnet/jit-contrib

tannergooding commented on src/coreclr/jit/emit.cpp:
```diff
@@ -6330,11 +6330,9 @@ unsigned emitter::emitEndCodeGen(Compiler* comp,
-        UNATIVE_OFFSET roDataAlignment = TARGET_POINTER_SIZE; // 8 Byte align by default.
-        roDataAlignmentDelta = (UNATIVE_OFFSET)ALIGN_UP(emitTotalHotCodeSize, roDataAlignment) - emitTotalHotCodeSize;
-        assert((roDataAlignmentDelta == 0) || (roDataAlignmentDelta == 4));
+        roDataAlignmentDelta = AlignmentPad(emitTotalHotCodeSize, emitConsDsc.alignment);
```
Should we assert that `alignment` is at least `MinAlignment` (`4` in the case of Arm64?)

BruceForstall commented on src/coreclr/jit/emit.cpp:
```diff
@@ -6330,11 +6330,9 @@ unsigned emitter::emitEndCodeGen(Compiler* comp,
     }
 
     UNATIVE_OFFSET roDataAlignmentDelta = 0;
```
It's reasonable to reconsider. However, on arm64, we have limited addressing mode range for data load instructions. If we put the data in a "data section", we would either have to (1) generate pessimistic code to allow the largest possible range, (2) ensure that data section is "close enough" to the code, or (3) optimistically assume the data is "close enough" to the code, and allow a back-off/retry if not.

tannergooding replied:
Perhaps @TamarChristinaArm or our other friends at ARM could provide input here on what's the recommended/optimal approach and if Arm64 has similar considerations around having data/instructions close together.

TamarChristinaArm replied:
> I would expect Arm64 to have similar limitations and for us to likewise want this data separate from the code. This also includes for other reasons like preventing users from trying to execute "data", etc.

Indeed we do have similar issues on Arm64 and the NX bits are of particular interest these days.  What we try to do in these cases is to create an anchor to the data section, and then subsequent loads just use offsets from the anchor.

typically we also then consider the anchors cheap to re-materialize to avoid spilling them around call sites etc.

If you're doing NX bits you'd have to allocate new pages for the constants anyway, you could consider getting a page near the code. If you're within the range of an `adrp+add` you can use the `adrp` as the anchor.

tannergooding replied:
Thanks for the input here! I'll log an issue capturing this and ensuring we consider the potential impact longer term.

tannergooding replied:
Logged https://github.com/dotnet/runtime/issues/71155

BruceForstall committed:
Additional fixes

1. On arm64/LA64, if asking for a data alignment greater than code
aligment, we need to increase the requested code alignment since
the code section is where this data will live. This isn't viewable
in SPMI diffs, but it does increase the alignment of some functions
from 8 to 16 byte code alignment.
2. Assert that the data section is at least 4 bytes aligned
(this is the default in our code, and alignment only increases).
3. Simplify the code setting the alignment flags for allocMem.

BruceForstall commented:
Pushed a few improvements:

1. On arm64/LA64, if asking for a data alignment greater than code
alignment, we need to increase the requested code alignment since
the code section is where this data will live. This isn't viewable
in SPMI diffs, but it does increase the alignment of some functions
from 8 to 16 byte code alignment.
2. Assert that the data section is at least 4 bytes aligned
(this is the default in our code, and alignment only increases).
3. Simplify the code setting the alignment flags for allocMem.

kunalspathak reviewed (approved):
Code looks better now. Thanks!

BruceForstall committed:
Formatting + disable alignment asserts

It looks like the buffer pointer passed back from crossgen2 doesn't
meet the alignment request. Perhaps it does in the final image, but
not in the buffer the JIT fills in? Maybe the asserts could be used
for JIT-time but not AOT (when the buffer address is the final location
of the code/data)?

tannergooding commented on src/coreclr/jit/emit.cpp:
```diff
@@ -6299,23 +6304,44 @@ unsigned emitter::emitEndCodeGen(Compiler* comp,
+    {
+        allocMemFlagDataAlign = CORJIT_ALLOCMEM_FLG_RODATA_16BYTE_ALIGN;
+    }
+    else if (dataAlignment == 32)
```
Just wondering, is there a reason why this one is `16, 32` while `code` is `32, 16` (just order of the checks)?

What is the default for `data alignment` if nothing is specified, is it `4` or `8`?

BruceForstall replied:
> Just wondering, is there a reason why this one is `16, 32` while `code` is `32, 16` (just order of the checks)?

No reason; probably should have been more consistent, but it doesn't actually matter.

> What is the default for `data alignment` if nothing is specified, is it `4` or `8`?

It's 8, or 4 for 32-bit platforms with <8 bytes of data.

BruceForstall commented on src/coreclr/jit/emit.cpp:
```diff
@@ -6280,14 +6285,14 @@ unsigned emitter::emitEndCodeGen(Compiler* comp,
         if (emitComp->fgCalledCount > (scenarioHotWeight * emitComp->fgProfileRunsCount()))
         {
-            allocMemFlag = CORJIT_ALLOCMEM_FLG_16BYTE_ALIGN;
+            codeAlignment = 16;
```
I don't know. Maybe historical and should be revisited?

BruceForstall commented on src/coreclr/jit/emit.cpp:
```diff
@@ -6375,6 +6401,18 @@ unsigned emitter::emitEndCodeGen(Compiler* comp,
     {
         assert(((size_t)codeBlock & 31) == 0);
     }
+    if ((allocMemFlag & CORJIT_ALLOCMEM_FLG_16BYTE_ALIGN) != 0)
```
I suppose that makes sense. Probably should be some asserts places (including in the VM and crossgen) that doesn't allow setting multiple; I don't see those asserts today (at least they check 32 before 16).
